### PR TITLE
Logsignature example

### PR DESCRIPTION
**Pull request by jambo6**
This PR adds a simple logsignature example that broadly follows the spiral example but tests over a large step size for logsignature depths 1, 2, 3. The model achieves top performance using the depth 2 logsignature and thus gives an indication into the potential usefulness. 

This example computes logsignatures upfront as a preprocessing step and as such provides a very basic usage whereby one cannot easily query for values in between logsignature endpoints. My preferred usage is to build a dataset class that when queried with an interval returns a logsignature, the `signatory.Path` class is initialised in the dataset so we get fast retrieval of the logsignatures. In the training loop one can then provide a custom sampler with a collection of intervals that can be iterated over. What is nice about this is we can at any point query with a different interval (or collection thereof) to evaluate over different logsignatures. The code would look something like this
```
dataset = LogSignatureDataset(data, depth=3)
sampler = IntervalSampler(step=10, method='fixed_length_intervals')    # We can then also define a sampler where intervals are randomised in the training loop
dataloader = Dataloader(dataset, sampler=sampler)
model = NCDE(...)
for batch in dataloader:
    # Train model
# Evaluate over a new interval
new_intervals = [[0, 3], [3, 7], [7, 11]]
outputs = model(dataset[new_intervals])    # We now have the results over different times
```
I think this is a more natural method for this approach, although it seems a little complex. I'm not sure it would fit in the main code base, but would you be interested in having such an example given?

### CODE REVIEW

File: example/logsignature_example.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
+######################
+# Here's how to code up a Neural CDE using the log-ode method for a long time series.
+# This method assumes familiarity with the standard Neural CDE example at `example.py`. We will only describe the
```
Comment (patrick-kidger) [inline example/logsignature_example.py:3]: `example.py` has been renamed.

Can you include a title + arxiv reference for the NRDE paper here?

File: example/logsignature_example.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+# differences from that example.
+######################
+import time
+import torch
+import torchcde
+from time_series_classification import NeuralCDE, get_data
+
+torch.manual_seed(0)
```
Comment (patrick-kidger) [inline example/logsignature_example.py:11]: not really necessary in a simple example?

File: example/logsignature_example.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+
+def _train(train_X, train_y, test_X, test_y, depth=None, num_epochs=10, window_length=100):
```
Comment (patrick-kidger) [inline example/logsignature_example.py:14]: can optional arguments be avoided if they're going to be passed?

File: example/logsignature_example.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    return proportion_correct, elapsed
+
+
+def print_heading(message):
+    # Print a message inbetween rows of #'s
+    string_sep = "#" * 50
+    print("\n" + string_sep + "\n{}\n".format(message) + string_sep)
+
+
+def main(num_epochs=15):
```
Comment (patrick-kidger) [inline example/logsignature_example.py:62]: Is this a sensible `num_epochs` to get good results? It feels quite small.

File: example/logsignature_example.py [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    ######################
+    num_timepoints = 5000
+    train_X, train_y = get_data(num_timepoints=num_timepoints)
+    test_X, test_y = get_data(num_timepoints=num_timepoints)
+
+    ######################
+    # We test the model over logsignature depths [1, 2, 3] with a window length of 50. This reduces the effective
+    # length of the path to just 100. The only change is an application of `torchcde.logsig_windows`
+
+    # The number of input channels changes from 3 to [3, 6, 14] for depths [1, 2, 3] respectively. Higher depths contain
```
Comment (patrick-kidger) [inline example/logsignature_example.py:76]: the "3 to [3, 6, 14] for depths [1, 2, 3]" could be worded a little more clearly.

File: example/logsignature_example.py [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+    # more information about the path over the interval, at a cost of increased numbers of channels.
+    ######################
+    depths = [1, 2, 3]
+    window_length = 500
```
Comment (patrick-kidger) [inline example/logsignature_example.py:80]: should this be 50?

File: example/logsignature_example.py [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+        accuracies.append(acc)
+
+    # Finally log the results to the console for a comparison
+    print_heading("Final results")
+    for acc, elapsed, depth in zip(accuracies, training_times, depths):
+        print(
+            "Depth: {}\n\tAccuracy on test set: {:.1f}%\n\tTime per epoch: {:.1f}s".format(
+                depth, acc * 100, elapsed / num_epochs
+            )
+        )
```
Comment (patrick-kidger) [inline example/logsignature_example.py:98]: what's a typical printout here?

**Review by patrick-kidger — COMMENTED** [review]
Glad to see this going in. Happy to merge once the comments are addressed.

**Comment by patrick-kidger** [comment]
I think a general sampler is probably a step too complicated to include as a pre-built example, unless it really demonstrably improves performance.

File: example/logsignature_example.py (hunk 4 above)
Comment (jambo6) [reply example/logsignature_example.py:62]: depth 1 gets 78% accuracy, depth 2 gets 98% accuracy, at least for `torch.manual_seed(0)`.

File: example/logsignature_example.py (hunk 6 above)
Comment (jambo6) [reply example/logsignature_example.py:80]: Either works, I used 500 because it still gets nearly 100% accuracy at depth 2 and obviously runs a lot faster.

File: example/logsignature_example.py (hunk 7 above)
Comment (jambo6) [reply example/logsignature_example.py:98]:
```
##################################################
Final results
##################################################
Depth: 1
	Accuracy on test set: 78.6%
	Time per epoch: 7.3s
Depth: 2
	Accuracy on test set: 98.8%
	Time per epoch: 7.5s
Depth: 3
	Accuracy on test set: 79.2%
	Time per epoch: 7.9s
```

File: example/logsignature_example.py (hunk 6 above)
Comment (patrick-kidger) [reply example/logsignature_example.py:80]: Okay. I think I'd prefer 50 just for consistency with the other example. Either way either this or the comment above it needs updating.

Comment (jambo6) [reply example/logsignature_example.py:80]: Oh yes, I see, my mistake.
